### ISSUE TITLE
Is the "gate_proj" used?

### ISSUE THREAD

**Issue by billsun9** [allenai/OLMoE#37]
OlmoeForCausalLM(
  (model): OlmoeModel(
    (embed_tokens): Embedding(50304, 2048, padding_idx=1)
    (layers): ModuleList(
      (0-15): 16 x OlmoeDecoderLayer(
        (self_attn): OlmoeSdpaAttention(
          (q_proj): Linear(in_features=2048, out_features=2048, bias=False)
          (k_proj): Linear(in_features=2048, out_features=2048, bias=False)
          (v_proj): Linear(in_features=2048, out_features=2048, bias=False)
          (o_proj): Linear(in_features=2048, out_features=2048, bias=False)
          (q_norm): OlmoeRMSNorm((2048,), eps=1e-05)
          (k_norm): OlmoeRMSNorm((2048,), eps=1e-05)
        )
        (mlp): OlmoeSparseMoeBlock(
          (gate): Linear(in_features=2048, out_features=64, bias=False)
          (experts): ModuleList(
            (0-63): 64 x OlmoeMLP(
              **_(gate_proj): Linear(in_features=2048, out_features=1024, bias=False)_**
              (up_proj): Linear(in_features=2048, out_features=1024, bias=False)
              (down_proj): Linear(in_features=1024, out_features=2048, bias=False)
              (act_fn): SiLU()
            )
          )
        )
        (input_layernorm): OlmoeRMSNorm((2048,), eps=1e-05)
        (post_attention_layernorm): OlmoeRMSNorm((2048,), eps=1e-05)
      )
    )
    (norm): OlmoeRMSNorm((2048,), eps=1e-05)
    (rotary_emb): OlmoeRotaryEmbedding()
  )
  (lm_head): Linear(in_features=2048, out_features=50304, bias=False)
)

Hi @Muennighoff I'm curious if the gate_proj layer is used here. My thought process was for a single token position at some layer (I denote the architecture as []):

prerouting logits x in R^2048 --> [gate] --> routing logits in R^64 --> [softmax then top8] --> vector with 56 0's and the top8 indices have the softmax weights of the routing logits

the same prerouting logits x in R^2048 --> [up_proj] --> [SiLU] --> [down_proj] --> expert_i's output in R^2048

And the final output is a weighted sum of the above two terms. If so, what's the gate_proj doing?

**Comment by Muennighoff** [comment]
yes i think it is used in the SwiGLU

**Comment by billsun9** [comment]
Gotcha - so the gate_proj and SiLU components are both part of swiglu activation FN? I.e. for some expert, its like x --> [up_proj] --> (gate_proj + SiLU activation fn) --> [down_proj]?

```

def forward(self, x):
    gate_out = self.gate_proj(x)        # First half
    up_out = self.up_proj(x)            # Second half
    return self.down_proj(F.silu(gate_out) * up_out)

```

**Comment by billsun9** [comment]
Thanks for the help!
